### ISSUE TITLE
Startup/stop time improvement

### ISSUE THREAD

**Issue by vnagara** [fail2ban/fail2ban#3088]
#### Feature request type
Improve startup and stop time of fail2ban service (start / stop actions) by bulk ban / unban operations.

#### Description
Recently I found out that my server was slowing down because of `iptables` 20K rules. In digging out solutions I noticed that 
starting service with my 20K bans takes around 4 HOURS on 2CPU machine. That is a bit ridicules and frustrating on rebooting. With switching to `nftables` it seems time reduced to 30-40 mins (didn't do thorough checks). Stopping the service is the same long, never-ending process. However, using nftables config file takes less than 1sec to load all of 20K IPs (that how I'm gonna to speed up my fail2ban now, by transferring permanent (5-10 years) bans to nftables config file).

Here the another solution for that: to do a bulk ban / unban. By this I mean instead of ban 1 IP at a time do up to 5000 bans for a jail at service start therefore you will need just couple junks for 20K records. The same goes for stop service (actually, here we can just flush (nftables flush set / iptables -F).
Forgot to mention It will also reduce CPU load thus speed up servers after reboots.

#### Possible solution
To add variable <ips> on actionstart / actionstop invocation so we can add insert all IPs at start/stop of jail.
For example for **actions.d/nftables-commons.conf**: 
```
actionstart = <nftables> add set <nftables_family> <nftables_table> <set_name> \{ type <nftables_type>\; \}
              <nftables> insert rule <nftables_family> <nftables_table> <chain> %(nftables_mode)s <address_family> saddr @<set_name> <blocktype>
              #new row
              <nftables> add element <nftables_family> <nftables_table> <set_name> \{ <ips> \}
```
Also, skip whole unban actions on stop service as all records (a rule and a set) will be deleted/flushed with
```
actionstop = .....
             <nftables> delete rule <nftables_family> <nftables_table> <chain> $HANDLE_ID
             <nftables> delete set <nftables_family> <nftables_table> <set_name>
``` 
so you don't need waste time / CPU on unbaning.

Thanks for you attention.

**Comment by sebres** [comment]
I have an experimental branch with `actionbulkban` which allows to ban multiple IPs with a single command.
I must still find time to rebase it in stock fail2ban, provide several tests and extend several base actions with this functionality.
Unfortunately I'm persistently busy and don't get enough time to fulfill that. Please, be patient.

> ... iptables 20K rules. In digging out solutions I noticed that starting the service with my 20K bans takes around 4 HOURS on 2CPU machine

Well, you could try to switch to iptables-ipset action (which is substantial faster than iptables). Also newest version (1.0) will do it still faster due to merge of solution changing actioncheck behavior (see https://github.com/fail2ban/fail2ban/issues/488#issuecomment-574676768).

But 4h for 20K is anyway too heavy in my opinion (if I'm not wrong, it'd take few minutes with newest fail2ban).
Which fail2ban version exactly do you use?

> Possible solution
> actionstart = ...
> actionstop = ...

You don't need to extend `actionstop`, because fail2ban since v.0.10 already partially implements "bulk" unban (for all tickets) - it's called `actionflush` (if defined it should remove all tickets at once, and then the stop of jail/action would not perform individual unbanning for every single IP for this action).
As for `actionstart`, I think afore-mentioned `actionbulkban` will be exactly what you are searching for, just it can be used not at start only (but to reban IPs, to ban multiple IPs at once, etc).

**Comment by vnagara** [comment]
>I have an experimental branch with actionbulkban which allows to ban multiple IPs with a single command.

That's awesome. Should solve the problem.
>Please, be patient.

I am, I wished to share my idea before I transfered bans to nftables custom config (that was inspire by https://github.com/ritsu/ipset-fail2ban script) and forget about it for 1-2 years.
> Well, you could try to switch to iptables-ipset action (which is substantial faster than iptables).

Yes I could but opted for recommended by many distros substitute __nftables__ with `set` that already integrated in fail2ban package action.d/nftables-common.
> 
> But 4h for 20K is anyway too heavy in my opinion (if I'm not wrong, it'd take few minutes with newest fail2ban).

The significant slow down comes from `iptables` as adding new rule to 10K rules dramatically longer then to 1K rules and adds up. I believe ip-set would reduce this time similar to _nftables set_ (to 30-40 mins, still longer than 1 sec).
> Which fail2ban version exactly do you use?

I am using v.0.10.2

P.S. For `actionbulkban` you will need limit quantity of IP's to fit the 'bash max arg size' limit (in my case 20k didn't work, I split it by 5k)

P.S.S @sebres  Thank you for your efforts.

**Comment by sebres** [comment]
As for the limits, it depends on how it'd be supplied (fail2ban rather uses rather env-vars named `$f2bV_xxx`, but they have similar length restriction)... in my case I used pipe containing IDs separated by configured char (default space), but since it is still experimental, so possibly I'd redefine it yet (or make it configurable).

**Comment by brianjmurrell** [comment]
3.5 years later, wondering if there is any progress on this bulk [un-]ban concept.

**Comment by vnagara** [comment]
> 3.5 years later, wondering if there is any progress on this bulk [un-]ban concept.
I preload all 20k ips on startup of the system via /etc/nftables/blacklist.rules config.
So on start I have only latest ips.

**Comment by skibbipl** [comment]
Yeah, I would also love to see this feature: I've switched recently to mikrotik integration and banning hundreds of IPs during startup of fail2ban executes crazy amount of logins/disconnects to mikrotik.

**Comment by sgothel** [comment]
I concur w/ the importance of a bulk atomic nftable restore feature.

I also became victim of the [Haskell/Lighpanda incident](https://jausoft.com/cgit/jaunotes.git/about/config/server/xx-attack/DoS-Mitigation.md#haskelllightpanda-incident) @ jogamp.org and jausoft. Also [see here](https://mailman.haskell.org/archives/list/ghc-devs@haskell.org/thread/AKWY3G76BMMOS6CNV5PZ64PHNWGDK3MM/).

The referenced little summary, shows how I use firewall + nftables + fail2ban/apache w/ current badbots detection filter
leading to around 550k banned IP entries. A line by line nftable operation takes hours, if not days.

Hence IMO a nftable text file for the huge set should be generated offline and fed into nftable atomical at once.
E.g. dumping the huge table at once still takes some times (~15s or so) .. but it is bearable and working.

In my monitor (top alike) script, I also had to stop querying banned IPs via nftable (or fail2ban-client banned)
and query the [DB in readonly mode directly](https://jausoft.com/cgit/jaunotes.git/tree/config/server/xx-attack/scripts/print-network-stats.sh#n64).

Last but not least - and most importantly, thank you all and @sebres for your great work w/ fail2ban!

**Comment by sgothel** [comment]
> Hence IMO a nftable text file for the huge set should be generated offline and fed into nftable atomical at once. E.g. dumping the huge table at once still takes some times (~15s or so) .. but it is bearable and working.
> 

@sebres if this 'job' is open and 
- you are not intending to working on it (but may provide your mentioned work so far)
- and would accept this feature (nftable only)
I would volunteer to create a PR, since I could use it to speed up restore.

`norestore` is not an option for me, since I want to preserve the bans and IP information
to ease server load - not even allowing to let them hit any port and service for a long while.
The latter should be the fair costs of abusing our scarce resources (the AI scraper).
